### ISSUE TITLE
stable diffusion settings

### ISSUE THREAD

**Issue by chrisrude** [chrisrude/oobabot-plugin#2]
Add a UI element for SD request params.

**Comment by sidonsoft** [comment]
I have setup the extension today, can you also please elaborate on the stable diffusion prefix and provide an example of use of the prefix.  Does the extension save settings create a config file that can be edited with settings that the extension UI doesn't provide such as resolution, sample, steps etc.  How does it handle img2ime and user uploaded images?

**Comment by chrisrude** [comment]
great questions!  Let me put in a few small tweaks and see if they answer your questions.

Please understand that I'm not trying to be obtuse, these are just great questions, and I know that if you have these questions than other users will too.  So the challenge is how do I update the interface so that it's more clear what's going on, so that folks can do what they want right away.  Stay tuned!

**Comment by sidonsoft** [comment]
originally I had to edit and build from source, then the config file update and made that process alot easier but I still didnt understand how it generated the prompt for the stable diffusion, now that it is an extension.  I am trying to understand how the extension gets the parameters and also curious how the regex update works

**Comment by chrisrude** [comment]
Version 0.1.9 is out!  It should address a number of the questions you had above.  Let me know what you think!

**Comment by sidonsoft** [comment]
exposing the config yaml is a great idea for advanced users, are we able to add stable diffusion request parameters into the config file such as CFG scale etc as per api json schema? 

can you elaborate on the user_ai_generated_keywords and regex method?

**Comment by chrisrude** [comment]
both of these are features that I haven't really had time to flesh out.

Normally, keywords for a stable diffusion request are just extracted straight from a message.  e.g. if the chat user posts
> make me a photo of dogs playing poker
the string "dogs playing poker" is what's sent to SD

this works pretty well once users get the idea of how to prompt, but works less well if the sentence is less well-structured, e.g.
> I'd like to see dogs playing poker, could you make me a photo of it?
the only thing that would get sent is "of it", and the image would be bogus

with `user_ai_generated_keywords` on, what happens is a little different.  Instead of splitting up the string and sending parts of it to SD, we instead take the whole string, add in a different prompt, and send that off to Oobabooga first.  The prompt we include is something like:
> "This message includes a request to generate an image.  please return a list of keywords for that image"

In the good csae, Oobabooga would respond with something like "keywords: [dogs, poker, table, cards, artwork]".  This is what it would send to SD.

It worked reasonably well on models which were trained for instruction-following, but the ones that were optimized for chat were really bad at it.  They'd try to talk about the idea of making photos, and have no idea what we were asking for.

So that's why it's partially done, I'd need some method to know more about what model is at the other end and how to prompt it to give reliable results.

**Comment by chrisrude** [comment]
Regex method is a lot more straightforward, and works, but requires that users understand regex, which super tech savvy folks do, but I'm aiming this at a more general audience.  It's for the models that are trained to follow a very specific chat pattern, e.g.:
```
##RESPONSE: (chat message1).  <-message-> (chat message 2) <-message->
```
(just making something up here, but the pattern is the same.

With the regex feature the user could write a regular expression to say exactly how this message should be split, e.g.
   "^[RESPONSE: ]?([^\<]+)<-message->"
and just the stuff in the first capture group (the parentheses above) would be sent.

This is half-done since it's really hard to debug if you get it wrong, and I suspect that in a week or two models will have standardized on a handful of response formats, making this not that useful long-term.

**Comment by bilionz** [comment]
Is there any way to pass SD api request params as custom Defaults from the bot configs ?
I m reffering to the /sdapi/v1/txt2img POST from the SD api.

**Comment by chrisrude** [comment]
Yes!  Open your `config.yml` file (it should be in the same directory as wherever you're running `oobabot` from.  There's a section most of the way down called `stable_diffusion`, and then within that another section called `request_params`.  It should look something like this:

```yaml
  request_params:
    do_not_save_samples: true
    do_not_save_grid: true
    negative_prompt: animal harm, suicide, loli, nsfw
    negative_prompt_nsfw: animal harm, suicide, loli
    steps: 30
    width: 512
    height: 512
    sampler: ""
```

Whatever you put there will be converted to .json and included in the POST to `/sdapi/v1/text2img`.

Was there something in particular you were interested in changing?

Let me know if this works for you, or if you have any more questions.
